### PR TITLE
S9: Xen 4.12 with v4v

### PR DESCRIPTION
**Pull request by tsirakisn**
Sorry for the duplicate PRs. This PR updates xen from 4.11 to 4.12 and temporarily re-introduces v4v while argo is sorted out.

Note: all changes are specific to 4.12 uprev

Merge with:
https://github.com/OpenXT/xsm-policy/pull/29
https://github.com/OpenXT/v4v/pull/40

OXT-1454
OXT-1604

### CODE REVIEW

**Comment by rneilturner** [comment]
Recommend updating the XEN_SRCREV to 714207b3fae7498831a0df338e6a1fbfeea594a7j [stable-4.12], brining in the spectre fixes by Andrew Cooper. These changes will depend on the PR's for microcode and Linux Kernel I posted earlier.

**Comment by crogers1** [comment]
That would also address https://openxt.atlassian.net/browse/OXT-1604

**Comment by eric-ch** [comment]
Build [$6563](http://openxt-builder.ainfosec.com:8010/builders/openxt64/builds/6563)

**Comment by stacktrust** [comment]
Please add OXT-1604 (security fixes) and OXT-1454 (Xen 4.12 uprev) to the PR description

File: recipes-extended/xen/files/hvmloader-legacy-seabios-optionroms.patch [hunk 1]
```diff
@@ -122,10 +140,28 @@ PATCHES
  struct seabios_info {
      char signature[14]; /* XenHVMSeaBIOS\0 */
      uint8_t length;     /* Length of this struct */
-@@ -141,10 +153,66 @@ static void seabios_load(const struct bi
-     seabios_config.image_size = bios_length;
+@@ -130,9 +142,11 @@ static void seabios_setup_e820(void)
+     dump_e820_table(e820, info->e820_nr);
  }
  
++static void *ipxe_module_addr;
```
Comment (stacktrust) [inline recipes-extended/xen/files/hvmloader-legacy-seabios-optionroms.patch]: Is there a ticket number for the PXE related changes?  Or is this related to Xen 4.12 uprev?  Can this change go upstream?

File: recipes-extended/xen/files/libxl-RFC-5of7-Handle-Linux-stubdomain-specific-QEMUoptions.patch [hunk 2]
```diff
@@ -64,23 +64,40 @@ PATCHES
...
+@@ -1157,38 +1161,18 @@ static int libxl__build_device_model_arg
                        "-xen-domid",
                        GCSPRINTF("%d", guest_domid), NULL);
  
 -    flexarray_append(dm_args, "-chardev");
--    flexarray_append(dm_args,
--                     GCSPRINTF("socket,id=libxl-cmd,"
--                                    "path=%s/qmp-libxl-%d,server,nowait",
--                                    libxl__run_dir_path(), guest_domid));
+-    if (state->dm_monitor_fd >= 0) {
```
Comment (stacktrust) [inline recipes-extended/xen/files/libxl-RFC-5of7-Handle-Linux-stubdomain-specific-QEMUoptions.patch]: Why were these changes needed?  Can we use any of the work done by Qubes with upstream Xen 4.12 for Linux stubdoms?

File: recipes-extended/xen/files/libxl-openxt-helpers.patch [hunk 3]
```diff
@@ -113,51 +113,79 @@ PATCHES
      libxl__store_libxl_entry(gc, dm_domid, "stubdom-version",
 --- a/tools/libxl/libxl_domain.c
 +++ b/tools/libxl/libxl_domain.c
-@@ -1023,6 +1023,7 @@ void libxl__destroy_domid(libxl__egc *eg
+@@ -1020,11 +1020,15 @@ static void domain_destroy_domid_cb(libx
+                                     libxl__ev_child *destroyer,
+                                     pid_t pid, int status);
+ 
++/* OpenXT functions */
```
Comment (stacktrust) [inline recipes-extended/xen/files/libxl-openxt-helpers.patch]: Would be helpful to have a comment in the PR description, or a separate JIRA ticket, describing why these changes were necessary.  Are they specific to the 4.12 uprev, or fixing a general issue?

File: recipes-extended/xen/files/libxl-vwif-support.patch [hunk 4]
```diff
@@ -319,17 +319,17 @@ PATCHES
      ("coloft_forwarddev", string),
 --- a/tools/libxl/libxl_types_internal.idl
 +++ b/tools/libxl/libxl_types_internal.idl
-@@ -29,6 +29,7 @@ libxl__device_kind = Enumeration("device
-     (12, "VDISPL"),
+@@ -30,6 +30,7 @@ libxl__device_kind = Enumeration("device
      (13, "VUART"),
      (14, "PVCALLS"),
-+    (15, "VWIF"),
+     (15, "VSND"),
```
Comment (stacktrust) [inline recipes-extended/xen/files/libxl-vwif-support.patch]: Ticket number for this change?

File: recipes-extended/xen/files/tboot-measure-and-launch-from-xen-efi.patch [hunk 5]
```diff
@@ -900,6 +898,8 @@ Signed-off-by: Tamas K Lengyel <lengyelt@ainfosec.com>
 -    pa += sizeof(uint64_t);
 -    tboot_copy_memory((unsigned char *)&sinit_mle_data, sizeof(sinit_mle_data),
 -                      pa);
++    /* map TXT heap into Xen addr space */
```
Comment (stacktrust) [inline recipes-extended/xen/files/tboot-measure-and-launch-from-xen-efi.patch]: Is this change specific to Xen 4.12 uprev or related to another ticket/issue?

File: recipes-extended/xen/files/hvmloader-legacy-seabios-optionroms.patch (hunk 1 above)
Comment (tsirakisn) [reply recipes-extended/xen/files/hvmloader-legacy-seabios-optionroms.patch]: See https://github.com/xen-project/xen/commit/e4506f404cb7983d6b424054ba999c6dd26ab4fa for the related commit. The IPXE module is no longer embedded (it is now loaded as an external module), so we can't load the etherboot ROM using our current methodology. Instead, we have to store the address of the external module at the time of the load and reference it later. I mirrored their changes for the ROMBIOS here.

I don't believe this change can go upstream as upstream doesn't support building/loading SEABIOS ROMs from out-of-tree (i.e. the original purpose of this patch).

File: recipes-extended/xen/files/libxl-RFC-5of7-Handle-Linux-stubdomain-specific-QEMUoptions.patch (hunk 2 above)
Comment (tsirakisn) [reply recipes-extended/xen/files/libxl-RFC-5of7-Handle-Linux-stubdomain-specific-QEMUoptions.patch]: It's hard to tell from this patch diff, but this was just a contextual adjustment. The end result is the same.

File: recipes-extended/xen/files/libxl-openxt-helpers.patch (hunk 3 above)
Comment (tsirakisn) [reply recipes-extended/xen/files/libxl-openxt-helpers.patch]: I should have added comments to all of these, will keep that in mind for the future. All of these changes are specific to the 4.12 uprev (added to PR description). For this diff in particular, it was another (mostly) contextual adjustment. I grouped the openxt helper bits in their own function, kill_openxt_helpers, and moved the call inside of a function that was added in 4.12. The changes should functionally be the same as before (kill the openxt helpers if the domain being destroyed has a stubdom).

File: recipes-extended/xen/files/libxl-vwif-support.patch (hunk 4 above)
Comment (tsirakisn) [reply recipes-extended/xen/files/libxl-vwif-support.patch]: The related commit can be found here: https://github.com/xen-project/xen/commit/0c798b22a009fd6e3ec388c11c1752e5eaa561aa

No change on our end, just a contextual adjustment to a change in the surrounding code.

File: recipes-extended/xen/files/tboot-measure-and-launch-from-xen-efi.patch (hunk 5 above)
Comment (tsirakisn) [reply recipes-extended/xen/files/tboot-measure-and-launch-from-xen-efi.patch]: Specific to 4.12. Another contextual adjustment.

File: recipes-extended/xen/files/libxl-openxt-helpers.patch [hunk 6]
```diff
@@ -113,51 +113,79 @@ PATCHES
...
-+                             GCSPRINTF("/local/domain/%d/"XS_QMP_PID, dm_domid));
-+        if (!pid)
-+            LOG(ERROR, "Cannot get qmp_helper pid for domain %d", dm_domid);
-+        else
-+            kill(strtol(pid, NULL, 10), SIGKILL);
+@@ -1107,6 +1117,34 @@ static void dm_destroy_cb(libxl__egc *eg
+     libxl__devices_destroy(egc, &dis->drs);
+ }
+ 
++static void kill_openxt_helpers(libxl__gc *gc, uint32_t dm_domid) {
```
Comment (jandryuk) [inline recipes-extended/xen/files/libxl-openxt-helpers.patch]: Can you move this function earlier to avoid the forward declaration?

File: recipes-extended/xen/files/hvmloader-legacy-seabios-optionroms.patch [hunk 7]
```diff
@@ -43,9 +43,9 @@ PATCHES
 +#SUBDIRS-$(CONFIG_SEABIOS) += seabios-dir
```
Comment (jandryuk) [inline recipes-extended/xen/files/hvmloader-legacy-seabios-optionroms.patch]: Since we run configure with '--with-system-seabios', CONFIG_SEABIOS is set to 'n' in `config/Tools.mk` and seabios-dir won't be included in SUBDIRS.  So this hunk could be dropped.

File: recipes-extended/xen/files/add-xc-hypercall-page.patch [hunk 8]
```diff
@@ -1,14 +1,12 @@
...
-@@ -59,8 +59,10 @@
- DECLARE_HYPERCALL(domctl)
- DECLARE_HYPERCALL(kexec_op)
+--- a/xen/arch/x86/guest/hypercall_page.S
++++ b/xen/arch/x86/guest/hypercall_page.S
+@@ -61,6 +61,9 @@ DECLARE_HYPERCALL(kexec_op)
  DECLARE_HYPERCALL(tmem_op)
--DECLARE_HYPERCALL(xc_reserved_op)
-+DECLARE_HYPERCALL(v4v_op)
+ DECLARE_HYPERCALL(argo_op)
```
Comment (stacktrust) [inline recipes-extended/xen/files/add-xc-hypercall-page.patch]: Will the Argo and v4v hypercalls collide?

File: recipes-extended/xen/files/v4v.patch [hunk 9]
```diff
@@ -2651,35 +2638,42 @@ PATCHES
...
  #define __HYPERVISOR_domctl               36
  #define __HYPERVISOR_kexec_op             37
  #define __HYPERVISOR_tmem_op              38
--#define __HYPERVISOR_xc_reserved_op       39 /* reserved for XenClient */
+-#define __HYPERVISOR_argo_op              39
 +#define __HYPERVISOR_v4v_op               39 /* reserved for XenClient */
  #define __HYPERVISOR_xenpmu_op            40
  #define __HYPERVISOR_dm_op                41
+ #define __HYPERVISOR_txt_op               42
++#define __HYPERVISOR_argo_op              43
```
Comment (dozylynx) [inline recipes-extended/xen/files/v4v.patch]: I don't think this change is the correct way to patch in v4v into Xen 4.12: it disables compatibility with Argo, which is the upstream technology. Instead, v4v should be issued the hypercall number 43, leaving Argo in place with 39. The numbering should be updated in the v4v Linux device driver to match the change here.

File: recipes-extended/xen/files/v4v.patch [hunk 10]
```diff
@@ -2651,35 +2638,42 @@ PATCHES
...
  #define VIRQ_CON_RING   8  /* G. (DOM0) Bytes received on console            */
  #define VIRQ_PCPU_STATE 9  /* G. (DOM0) PCPU state changed                   */
- #define VIRQ_MEM_EVENT  10 /* G. (DOM0) A memory event has occured           */
--#define VIRQ_XC_RESERVED 11 /* G. Reserved for XenClient                     */
+ #define VIRQ_MEM_EVENT  10 /* G. (DOM0) A memory event has occurred          */
+-#define VIRQ_ARGO       11 /* G. Argo interdomain message notification       */
 +#define VIRQ_V4V        11 /* G. Reserved for XenClient                     */
  #define VIRQ_ENOMEM     12 /* G. (DOM0) Low on heap memory       */
  #define VIRQ_XENPMU     13 /* V.  PMC interrupt                              */
++#define VIRQ_ARGO       14 /* G. Argo interdomain message notification       */
```
Comment (dozylynx) [inline recipes-extended/xen/files/v4v.patch]: Again, I don't think this is the correct way to patch v4v into Xen 4.12: v4v should be issued the VIRQ number 14, and leave Argo with its VIRQ number 11.

File: recipes-core/packagegroups/packagegroup-xenclient-dom0.bb [hunk 11]
```diff
@@ -16,7 +16,6 @@ RDEPENDS_${PN} = " \
     xen-hypervisor \
     xen-efi \
     xen-flask-tools \
-    xen-libxenctrl \
```
[the remark this answers is not in the document]
Comment (tsirakisn) [reply recipes-core/packagegroups/packagegroup-xenclient-dom0.bb]: No, xen-libxenctrl will still be installed. It is now an RPROVIDES of the xen package.

File: recipes-extended/xen/files/v4v.patch (hunk 9 above)
Comment (tsirakisn) [reply recipes-extended/xen/files/v4v.patch]: I actually had that originally, but I changed it to what I have now to avoid modifying other repos (e.g. v4v in your example). If that's the desired approach I'll change it back.

File: recipes-extended/xen/files/add-xc-hypercall-page.patch (hunk 8 above)
Comment (tsirakisn) [reply recipes-extended/xen/files/add-xc-hypercall-page.patch]: ^ That is correct, they can live alongside each other as long as their hypercall macro numbers are separate.

File: recipes-extended/xen/files/libxl-openxt-helpers.patch (hunk 6 above)
Comment (tsirakisn) [reply recipes-extended/xen/files/libxl-openxt-helpers.patch]: Sure

**Comment by tsirakisn** [comment]
> Recommend updating the XEN_SRCREV to 714207b3fae7498831a0df338e6a1fbfeea594a7j [stable-4.12], bringing in the spectre fixes by Andrew Cooper. These changes will depend on the PR's for microcode and Linux Kernel I posted earlier.

@rneilturner there have been 7-8 commits added after 714207b, should I just use the HEAD of stable-4.12?

**Comment by rneilturner** [comment]
> > Recommend updating the XEN_SRCREV to 714207b3fae7498831a0df338e6a1fbfeea594a7j [stable-4.12], bringing in the spectre fixes by Andrew Cooper. These changes will depend on the PR's for microcode and Linux Kernel I posted earlier.
> 
> @rneilturner there have been 7-8 commits added after 714207b, should I just use the HEAD of stable-4.12?

Yes I would, although none of those additional commits directly impact the spectre fixes. Having our build and patches track the latest stable release is preferable in my opinion.

**Comment by tsirakisn** [comment]
> Recommend updating the XEN_SRCREV to 714207b3fae7498831a0df338e6a1fbfeea594a7j [stable-4.12], bringing in the spectre fixes by Andrew Cooper. These changes will depend on the PR's for microcode and Linux Kernel I posted earlier.

@rneilturner there have been 7-8 commits added after 714207b, should I just use the HEAD of stable-4.12?

> Please add OXT-1604 (security fixes) and OXT-1454 (Xen 4.12 uprev) to the PR description

@stacktrust I should remove OXT-1550 correct? I must have thought that was the 4.12 uprev ticket at a quick glance.

File: recipes-extended/xen/files/v4v.patch (hunk 9 above)
Comment (dozylynx) [reply recipes-extended/xen/files/v4v.patch]: Yes, the change would be appreciated. Thanks.

File: recipes-core/packagegroups/packagegroup-xenclient-dom0.bb (hunk 11 above)
[the remark this answers is not in the document]
Comment (jandryuk) [reply recipes-core/packagegroups/packagegroup-xenclient-dom0.bb]: If actual package dependencies are set properly, I don't think there should be any `xen-lib*` entries in this RDEPENDS.  Installing `xen-xl` and the other xen packages should pull in whatever they need.

Did something in the 4.12 uprev require this change?

File: recipes-extended/xen/files/libxl-xenmgr-support.patch [hunk 12]
```diff
@@ -577,6 +579,18 @@ PATCHES
...
+ int libxl__qmp_restore(libxl__gc *gc, int domid, const char *state_file)
  {
      libxl__json_object *args = NULL;
+@@ -1047,6 +1052,11 @@ static int qmp_change(libxl__gc *gc, lib
+     return rc;
+ }
+ 
++int libxl__qmp_stop(libxl__gc *gc, int domid)
++{
++    return qmp_run_command(gc, domid, "stop", NULL, NULL, NULL);
```
[the remark this answers is not in the document]
Comment (tsirakisn) [reply recipes-extended/xen/files/libxl-xenmgr-support.patch]: If possible I'd like to get a second opinion on this; it looked okay to me, but I wasn't confident enough that there wouldn't be any undesired side effects.

Here's the original libxl__qmp_stop:
```
int libxl__qmp_stop(libxl__gc *gc, int domid)
{
    return qmp_run_command(gc, domid, "stop", NULL, NULL, NULL);
}
```

And here's the new suspend save:
```
void libxl__qmp_suspend_save(libxl__egc *egc,
                             libxl__domain_suspend_state *dsps)
{
    EGC_GC;
    int rc;
    libxl__ev_qmp *ev = &dsps->qmp;

    ev->ao = dsps->ao;
    ev->domid = dsps->domid;
    ev->callback = dm_stopped;
    ev->payload_fd = -1;

    rc = libxl__ev_qmp_send(gc, ev, "stop", NULL);
    if (rc)
        goto error;

    return;

error:
    dsps->callback_device_model_done(egc, dsps, rc);
}
```

[the remark this answers is not in the document]
Comment (jandryuk) [reply recipes-extended/xen/files/libxl-xenmgr-support.patch]: Re-introducing libxl__qmp_stop is probably the simplest change to keep the existing code working.  But it is re-introducing something that was removed upstream, so it will be maintenance going forward.

**Comment by stacktrust** [comment]
> > Please add OXT-1604 (security fixes) and OXT-1454 (Xen 4.12 uprev) to the PR description
> 
> @stacktrust I should remove OXT-1550 correct? I must have thought that was the 4.12 uprev ticket at a quick glance.

Yes please, that was a ticket for upstream OE changes

**Comment by jean-edouard** [comment]
Built here: http://openxt-builder.ainfosec.com:8010/builders/openxt64/builds/6583

**Comment by jean-edouard** [comment]
LGTM.
@dozylynx have your comments been addressed?

**Comment by dozylynx** [comment]
Yes, my comments on this PR have been addressed. There's a few minor things that would be nice to change if the v4v patch were going to live any longer: the insertion points for v4v_op are not strictly correct wrt hypercall number order in the source code, though they will function fine, and there are tab whitespace characters that this change is preserving that would be nice to drop, but it looks OK to proceed.

File: recipes-core/packagegroups/packagegroup-xenclient-dom0.bb (hunk 11 above)
[the remark this answers is not in the document]
Comment (eric-ch) [reply recipes-core/packagegroups/packagegroup-xenclient-dom0.bb]: @jandryuk I don't think so, but removing all the `xen-lib*` packages will have the correct dependencies installed. Filed [OXT-1620](https://openxt.atlassian.net/browse/OXT-1620) to track this and another complementary efforts. 

Since this PR is blocking further development and that hunk has introduces no risk, I think this can be addressed later.

**Comment by jean-edouard** [comment]
Built here: http://openxt-builder.ainfosec.com:8010/builders/openxt64/builds/6593

**Comment by jean-edouard** [comment]
LGTM, merging soon.
